src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -3035,16 +3058,36 @@ internal Task WriteByteArray(byte[] b, int len, int offsetBuffer, bool canAccumu
                                 completion = new TaskCompletionSource<object>();
                                 task = completion.Task; // we only care about return from topmost call, so do not access Task property in other cases
                             }
-                            WriteByteArraySetupContinuation(b, len, completion, offset, packetTask);
+
+                            if (array == null)
+                            {
+                                byte[] tempArray = new byte[len];
+                                Span<byte> copyTempTo = tempArray.AsSpan();
+                                ReadOnlySpan<byte> copyTempFrom = b.Slice(0, len);
```

Comment:
Can't you just get rid of the call at 3048. It doesn't seem to be doing anything useful. And call b.Slice(remainder, len) like you were doing earlier, at this location? `if ( array == null) ` is the only location where you are going to get the rest of the content of `b`
Does this make sense ?

Reply:
No, there is no guarantee that the WritePacket call will return a task so it's possible to go around the loop without allocating the array. This was a problem that occurred with one of the SplitPacket tests that was pointed out on the original PR.

Reply:
OK, now I understand. In that case, what is the effect of just doing a `b = b.Slice(remainder, len) ` at 3048 and eliminate this slicing at 3066. Would that work? Then this line would read `ReadOnlySpan<byte> copyTempFrom = b;`